src/services/codefixes/importFixes.ts

```diff
@@ -647,19 +642,43 @@ namespace ts.codefix {
     }
 
     function getImportCodeActions(context: CodeFixContext): CodeAction[] {
-        const { context: importFixContext, isJsxNamespace } = convertToImportCodeFixContext(context);
         return context.errorCode === Diagnostics._0_refers_to_a_UMD_global_but_the_current_file_is_a_module_Consider_adding_an_import_instead.code
-            ? getActionsForUMDImport(importFixContext, isJsxNamespace)
-            : getActionsForNonUMDImport(importFixContext, context.program.getSourceFiles(), context.cancellationToken);
+            ? getActionsForUMDImport(context)
+            : getActionsForNonUMDImport(context);
     }
 
-    function getActionsForUMDImport(context: ImportCodeFixContext, isJsxNamespace: boolean): CodeAction[] {
-        const { checker, symbolName, symbolToken, compilerOptions } = context;
-        const symbol = skipAlias(
-            isJsxNamespace ? checker.resolveName(symbolName, /*location*/ undefined, SymbolFlags.Value, /*excludeGlobals*/ false) : checker.getSymbolAtLocation(symbolToken),
-            checker);
-        return getCodeActionsForImport([{ moduleSymbol: symbol, importKind: getUmdImportKind(compilerOptions) }], { ...context });
+    function getActionsForUMDImport(context: CodeFixContext): CodeAction[] {
+        const token = getTokenAtPosition(context.sourceFile, context.span.start, /*includeJsDocComment*/ false);
+        const checker = context.program.getTypeChecker();
+
+        let umdSymbol: Symbol | undefined;
+
+        if (isIdentifier(token)) {
+            // try the identifier to see if it is the umd symbol
+            umdSymbol = checker.getSymbolAtLocation(token);
+        }
+
+        if (!isUMDExportSymbol(umdSymbol)) {
+            // The error wasn't for the symbolAtLocation, it was for the JSX tag itself, which needs access to e.g. `React`.
+            const parent = token.parent;
+            const isNodeOpeningLikeElement = isJsxOpeningLikeElement(parent);
+            if ((isJsxOpeningLikeElement && (<JsxOpeningLikeElement>parent).tagName === token) || parent.kind === SyntaxKind.JsxOpeningFragment) {
+                umdSymbol = checker.resolveName(checker.getJsxNamespace(),
+                    isNodeOpeningLikeElement ? (<JsxOpeningLikeElement>parent).tagName : parent, SymbolFlags.Value, /*excludeGlobals*/ false);
+            }
+        }
+
+        if (isUMDExportSymbol(umdSymbol)) {
+            const symbol = checker.getAliasedSymbol(umdSymbol);
+            if (symbol) {
+                return getCodeActionsForImport([{ moduleSymbol: symbol, importKind: getUmdImportKind(context.program.getCompilerOptions()) }],
+                    convertToImportCodeFixContext(context, isIdentifier(token) ? token : undefined, umdSymbol.name));

```

Comment:
`context.symbolToken` is currently typed as `Node | undefined` and we check if it's an identifier when using it. So there is currently redundant checking. We could change its type to `Identifier | undefined` and remove the check at its use, or remove the check here.

Reply:
done.